### PR TITLE
Validate the ocular config before generating gatsby config

### PR DESCRIPTION
**Pull request by javidhsueh**
- Validate user's input config and print all warnings
 - Install validate.js as the main mechanism to validate input.
 - Add extensive tests on validate-config.js

### CODE REVIEW

File: modules/gatsby/test/utils/validate-config.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
+/* eslint-disable max-len */
```
Comment (ibgreen) [inline modules/gatsby/test/utils/validate-config.js]: Usually we would call this file `validate-config.spec.js`

File: modules/gatsby/src/utils/validate-config.js [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+//
+// [presence](http://validatejs.org/#validators-presence)
+// The presence validator validates that the value is defined.
+//
+// [url](http://validatejs.org/#validators-url)
+// The URL validator ensures that the input is a valid URL.
+//
+// [custom validator](http://validatejs.org/#custom-validator)
+// Create our own custom reusable validator.
+const constraints = {
```
[the remark this answers is not in the document]
Comment (javidhsueh) [reply modules/gatsby/src/utils/validate-config.js]: Yes, we could. But I'll need to modify the `prerequisite` validator (line 99) used at line 393. Then we can make it a pure schema json file.

File: modules/gatsby/src/gatsby-config/get-gatsby-config.js [hunk 3]
```diff
@@ -12,59 +46,38 @@ module.exports = function getGatsbyConfig(config) {
...
+    pathPrefix: paddedConfig.PATH_PREFIX,
 
     // Site Metadata is populated from config (and react-helmet, see gatsby-plugin-react-helmet)
     siteMetadata: {
-      config: {...config, ...paddedConfig},
+      config: paddedConfig,
 
-      siteUrl: urljoin(config.PROJECT_URL, config.PATH_PREFIX)
+      siteUrl: urljoin(paddedConfig.PROJECT_URL, paddedConfig.PATH_PREFIX)
       /*
```
[the remark this answers is not in the document]
Comment (javidhsueh) [reply modules/gatsby/src/gatsby-config/get-gatsby-config.js]: what do you mean the 'These lower case settings are from an old gatsby starter' ? `rssMetadata`

[the remark this answers is not in the document]
Comment (ibgreen) [reply modules/gatsby/src/gatsby-config/get-gatsby-config.js]: > what do you mean the 'These lower case setting 

Yes `rssMetadata` should be removed. I guess we do still use `siteUrl`.

File: modules/gatsby/src/index.js [hunk 4]
```diff
@@ -22,6 +22,9 @@ const {
   getReactComponent
 } = require('./utils/example-registry');
 
+const {validateConfig} = require('./utils/validate-config');
+
 module.exports.registerReactComponent = registerReactComponent;
```
[the remark this answers is not in the document]
Comment (javidhsueh) [reply modules/gatsby/src/index.js]: done.

**Comment by javidhsueh** [comment]
@ibgreen addressed all your comments above.

**Pull request merged**